cmd/utils.go

```diff
@@ -101,46 +101,72 @@ func MigrateDatabase(ctx context.Context, targetProfile profiles.TargetProfile,
-			if err = conversion.UpdateDDLForeignKeys(ctx, adminClient, dbURI, conv, ioHelper.Out); err != nil {
-				err = fmt.Errorf("can't perform update schema on db %s with foreign keys: %v", dbURI, err)
-				return bw, err
-			}
-		}
+		bw, err = migrateData(ctx, targetProfile, sourceProfile, ioHelper, conv, dbURI, adminClient, client, v)
 	case *SchemaAndDataCmd:
-		err = conversion.CreateOrUpdateDatabase(ctx, adminClient, dbURI, sourceProfile.Driver, targetProfile.TargetDb, conv, ioHelper.Out)
+		bw, err = migrateSchemaAndData(ctx, targetProfile, sourceProfile, ioHelper, conv, dbURI, adminClient, client, v)
+	}
+	if err != nil {
+		err = fmt.Errorf("can't migrate database: %v", err)
+		return nil, err
+	}
+	return bw, nil
+}
+
+func migrateSchema(ctx context.Context, targetProfile profiles.TargetProfile, sourceProfile profiles.SourceProfile,
+	ioHelper *utils.IOStreams, conv *internal.Conv, dbURI string, adminClient *database.DatabaseAdminClient) error {
+	err := conversion.CreateOrUpdateDatabase(ctx, adminClient, dbURI, sourceProfile.Driver, targetProfile.TargetDb, conv, ioHelper.Out)
+	if err != nil {
+		err = fmt.Errorf("can't create/update database: %v", err)
+		return err
+	}
+	return nil
+}
+
+func migrateData(ctx context.Context, targetProfile profiles.TargetProfile, sourceProfile profiles.SourceProfile,
+	ioHelper *utils.IOStreams, conv *internal.Conv, dbURI string, adminClient *database.DatabaseAdminClient, client *sp.Client, cmd *DataCmd) (*writer.BatchWriter, error) {
+	var (
+		bw  *writer.BatchWriter
+		err error
+	)
+	if !sourceProfile.UseTargetSchema() {
+		err = validateExistingDb(ctx, conv.TargetDb, dbURI, adminClient, client, conv)
 		if err != nil {
-			err = fmt.Errorf("can't create/update database: %v", err)
+			err = fmt.Errorf("error while validating existing database: %v", err)
 			return nil, err
 		}
-		bw, err = conversion.DataConv(ctx, sourceProfile, targetProfile, ioHelper, client, conv, true, v.WriteLimit)
-		if err != nil {
-			err = fmt.Errorf("can't finish data conversion for db %s: %v", dbURI, err)
-			return nil, err
+	}
+	bw, err = conversion.DataConv(ctx, sourceProfile, targetProfile, ioHelper, client, conv, true, cmd.WriteLimit)
+	if err != nil {
+		err = fmt.Errorf("can't finish data conversion for db %s: %v", dbURI, err)
+		return nil, err
+	}
+	if !cmd.SkipForeignKeys {
+		if err = conversion.UpdateDDLForeignKeys(ctx, adminClient, dbURI, conv, ioHelper.Out); err != nil {
+			err = fmt.Errorf("can't perform update schema on db %s with foreign keys: %v", dbURI, err)
+			return bw, err
 		}
-		if !v.SkipForeignKeys {
-			if err = conversion.UpdateDDLForeignKeys(ctx, adminClient, dbURI, conv, ioHelper.Out); err != nil {
-				err = fmt.Errorf("can't perform update schema on db %s with foreign keys: %v", dbURI, err)
-				return bw, err
-			}
+	}
+	return bw, nil
+}
+
+func migrateSchemaAndData(ctx context.Context, targetProfile profiles.TargetProfile, sourceProfile profiles.SourceProfile,
```

Comment:
basic question - Why is this step not just calling migrateSchema and migrateData ?

Reply:
because in migrateData we validate the existing db which is not the case in schemaAndData